tests/moduletool/test_add.py

```diff
@@ -0,0 +1,209 @@
+        with open(requirements_txt_file, "r", encoding="utf-8") as fd:
+            return fd.read()
+
+    def _assert_project_state(constraint: str) -> None:
+        with open(project.get_metadata_file_path(), "r", encoding="utf-8") as fd:
+            project_metadata = ProjectMetadata.parse(fd)
+            assert constraint in project_metadata.requires
+        with open(os.path.join(snippetcompiler_clean.libs, "std", ModuleV1.MODULE_FILE), "r", encoding="utf-8") as fd:
+            module_metadata = ModuleV1Metadata.parse(fd)
+            assert module_metadata.version == constraint.split("==")[1]
+        assert os.listdir(snippetcompiler_clean.libs) == ["std"]
+        assert not _get_content_requirements_txt_file()
+
+    assert not os.listdir(snippetcompiler_clean.libs)
+    assert not _get_content_requirements_txt_file()
+
+    version_constraint = "std==3.0.2"
+    ModuleTool().add(module_req=version_constraint, v1=True, override=False)
+    _assert_project_state(version_constraint)
+
+    new_version_constraint = "std==3.0.1"
+    with pytest.raises(CLIException, match="A dependency on the given module was already defined"):
+        ModuleTool().add(module_req=new_version_constraint, v1=True, override=False)
+    _assert_project_state(version_constraint)
+
+    ModuleTool().add(module_req=new_version_constraint, v1=True, override=True)
+    _assert_project_state(new_version_constraint)
+
+
+@pytest.mark.slowtest
+def test_module_add_v2_module_to_project(
+    tmpdir: py.path.local, snippetcompiler_clean, local_module_package_index: str, modules_v2_dir: str
+) -> None:
+    """
+    Add a V2 module to an inmanta project using the `inmanta module add` command.
+    """
+    # Ensure that version 1.1.1, 1.2.0 and 1.2.3 exist in the pip index
+    pip_index = PipIndex(artifact_dir=os.path.join(str(tmpdir), "pip-index"))
+    for version in ["1.1.1", "1.2.0"]:
+        module_from_template(
+            source_dir=os.path.join(modules_v2_dir, "elaboratev2module"),
+            dest_dir=os.path.join(tmpdir, f"elaboratev2module-v{version}"),
+            new_version=Version(version),
+            publish_index=pip_index,
+        )
+
+    # Create project
+    project: Project = snippetcompiler_clean.setup_for_snippet(
+        snippet="", autostd=False, python_package_sources=[local_module_package_index, pip_index.url]
+    )
+
+    requirements_txt_file = os.path.join(project.path, "requirements.txt")
+
+    def _assert_project_state(pkg_name: str, expected_version: Version, project_requires_constraint: str) -> None:
+        installed_packages = process_env.get_installed_packages()
+        assert pkg_name in installed_packages
+        assert installed_packages[pkg_name] == expected_version
+        with open(project.get_metadata_file_path(), "r", encoding="utf-8") as fd:
+            project_metadata = ProjectMetadata.parse(fd)
+            assert pkg_name[len("inmanta-module-") :] in project_metadata.requires
```

Comment:
If I understand correctly, for a v1 module any v2 dependencies are trimmed from the yaml and only listed in the requirements file while for a project this is not the case. Is there a reason for this? Shouldn't we at least make sure only one location contains a constraint?

Reply:
* A V2 dependency of a V1 module is only set the requirements.txt file, because it's a Python dependency and V1 modules store their Python dependencies in the requirements.txt file.
* A V2 dependency on a project is present in the requirements.txt and the project.yml file. I chose to only mention the module name in the project.yml file, while the requirements.txt file also contains the actual version constraint. We can also choose to not put the dependency in the project.yml file. @wouterdb What is you opinion on this?

Reply:
I think it's all right to put it in the `project.yml` as well, but if we do I think when we add a constraint to the requirements file we need to make sure there is no constraint in the project.yml and remove it if there is.